Contributing Guide: Add to Sphinx (#888)
ECP-WarpX/WarpX: CONTRIBUTING.rst

.. _developers-contributing:

Contribute to WarpX
===================

We welcome new contributors!
Here is how to participate to the WarpX development.

Git workflow
------------

The WarpX project uses `git <https://git-scm.com>`_ for version control.
If you are new to git, you can follow one of these tutorials:

- `Learn git with bitbucket <https://www.atlassian.com/git/tutorials/learn-git-with-bitbucket-cloud>`_
- `git - the simple guide <http://rogerdudler.github.io/git-guide/>`_

Make your own fork and create a branch on it
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The basic WarpX workflow is:

- Fork the main repo (or update it if you already created it).
- Implement your changes and push them on a new branch ``<branch_name>`` on your fork.

> - Create a Pull Request from branch ``<branch_name>`` on your fork to branch `master` on the main WarpX repo.

Reviewer: Not sure if too much information for real beginners, but I definitely think it is useful to include in the workflow:
```suggestion
- Run the scripts `.github/workflows/source/hasEOLwhiteSpace`; `.github/workflows/source/hasNonASCII` and `.github/workflows/source/hasTabs` (in the shell you can do that by typing `bash <name_of_script` inside the WarpX repo main folder)
- Use git to compare the changes you did in branch ``<branch_name>`` on your fork to branch `master` on the main WarpX repo
- Create a Pull Request.
```

Author: I am note sure to include this right now and here, we might rather do an extra section about local testing in the developer docs in a follow-up (style and unit tests).

> First, let us setup your local git repo. Make your own fork of the main (``upstream``) WarpX repo:

Reviewer: Maybe it is clearer if the list between lines 23 and 25 is numbered (or a sub-sections list) and then these steps fall into each of those numbers. I can be wrong about this, but I think this way it is not as much jumping forth and back as much.

Author: Sounds good, I'll add 1.-3. for now :)

on the `WarpX Github page <https://github.com/ECP-WarpX/WarpX>`_, press the fork button.
Then, you can execute:

.. code-block:: sh

   # These 4 first lines are the same as for a standard WarpX install
   mkdir warpx_directory
   cd warpx_directory
   git clone --branch master https://bitbucket.org/berkeleylab/picsar.git
   git clone --branch development https://github.com/AMReX-Codes/amrex.git

   # Clone your fork on your local computer. You can get this address on your fork's Github page.
   git clone https://github.com/<myGithubUsername>/WarpX.git
   cd warpx
   # Keep track of the main WarpX repo, to remain up-to-date.
   git remote add upstream https://github.com/ECP-WarpX/WarpX.git

Now you are free to play with your fork (for additional information, you can visit the
`Github fork help page <https://help.github.com/en/articles/fork-a-repo>`_).

.. note::

   You do not have to re-do the setup above every time.

>    nstead, in the future, you need to update the ``master`` branch on your fork with

Reviewer: ```suggestion
   Instead, in the future,  all you need is to update the ``master`` branch on your fork with:
```

   .. code-block:: sh

      git checkout master
      git pull upstream master

Make sure you are on WarpX ``master`` branch with

.. code-block:: sh

   git checkout master

in the WarpX directory.

Create a branch ``<branch_name>`` (the branch name should reflect the piece of code you want to add, like ``fix_spectral_solver``) with

.. code-block:: sh

   git checkout -b <branch_name>

and do the coding you want.

> It is probably a good time to look at the `AMReX documentation <https://amrex-codes.github.io/amrex/docs_html/>`_ and at the `AMReX Doxygen reference <https://ccse.lbl.gov/pub/AMReX_Docs/index.html>`_.

Reviewer: I would move this line to before, when you are not explaining the GitHub part in detail. Because it is really important.

Author: Both fine for me, I think the idea here was: get set up and ready to code, now check the code docs.

Add the files you work on to the git staging area with

.. code-block:: sh

   git add <file_I_created> <and_file_I_modified>

Commit & push your changes
^^^^^^^^^^^^^^^^^^^^^^^^^^

Periodically commit your changes with

.. code-block:: sh

   git commit -m "This is a 50-char description to explain my work"

The commit message (between quotation marks) is super important in order to follow the developments and identify bugs.

For the moment, commits are on your local repo only.
You can push them to your fork with

.. code-block:: sh

   git push -u origin <branch_name>

If you want to synchronize your branch with the ``master`` branch (this is useful when the ``master`` branch is being modified while you are working on ``<branch_name>``), you can use

.. code-block:: sh

>    git pull upstream master

Reviewer: I might have done a mistake in my own repo structure, but I sometimes need the bottom line if I wish to merge to master into my branches. Right now I am not sure why.
```suggestion
   git pull upstream master
   git merge upstream/master
```
The latter line can be necessary, if merging doesn't happen automatically after you use the first line.

Author: `git pull` implies a merge unless `--ff-only` is specified. I would potentially add `git fetch` but that's also included in pull.

Reviewer: Good point!

and fix any conflict that may occur.

Submit a Pull Request
^^^^^^^^^^^^^^^^^^^^^

A Pull Request (PR) is the way to efficiently visualize the changes you made and to propose your new feature/improvement/fix to the WarpX project.
Right after you push changes, a banner should appear on the Github page of your fork, with your ``<branch_name>``.

- Click on the ``compare & pull request`` button to prepare your PR.
- It is time to communicate your changes: write a title and a description for your PR.
  People who review your PR are happy to know

>   * what feature/fix you propose, and why
>   * how you made it (created a new class than inherits from...)
>   * and anything relevant to your PR (performance tests, images, *etc.*)

Reviewer: ```suggestion
  * what feature/fix you propose, and why
  * how you made it (files changed fro what reason, created a new class than inherits from...)
  * how you tested it and what was the output you got
  * and anything relevant to your PR (performance tests, images, *etc.*)
```

Author: would not recommend that one has to explain why one touched each file. people might start explaining "first I changed the header because that's were function signatures go." which is a bit too verbose.
maybe `added new/edited files, create a new class that ...` is probably good

- Press ``Create pull request``.
  Now you can navigate through your PR, which highlights the changes you made.

Please DO NOT write large Pull Requests, as they are very difficult and time-consuming to review.
As much as possible, split them into small targeted PRs.
For example, if find typos in the documentation open a pull request that only fixes typos.
If you want to fix a bug, make a small pull request that only fixes a bug.
If you want to implement a large feature, write helper functionality first, test it and submit those as a first pull request.

> If you want to implement a feature and are not too sure how to split it, just open an issue about your plans and ping other WarpX developers on it to chime in.

Reviewer: ```suggestion
If you want to implement a feature and are not too sure how to split it, just open an issue about your plans and ping other WarpX developers on it to chime in.
You can also use the GitHub project tab in your own repo to organize the work into separate tasks/PRs and share it with the WarpX community to get feedback.
```

Author: Adopted and added! :+1:

Even before your work is ready to merge, it can be convenient to create a PR (so you can use Github tools to visualize your changes).

> In this case, please put the ``[WIP]`` tag (for Work-In-Progress) at the beginning of the PR title.

Reviewer: ```suggestion
In this case, please put the ``[WIP]`` tag (for Work-In-Progress) at the beginning of the PR title.
Another tag you may want to use is `[mini]`, if your changes are very few lines and quick to review.
```

Include a test to your PR
"""""""""""""""""""""""""

A new feature is great, a **working** new feature is even better!
Please test your code and add your test to the automated test suite.
It's the way to protect your work from adventurous developers.

> Instructions are given in the :ref:`testing section <developers-testing>` of our `developer's documentation <https://warpx.readthedocs.io/en/latest/developers/testing.html>`_.

Reviewer: This is really great! I was using it today thanks to Luca's PRs too.

> Include documentation to your PR

Reviewer: Sorry, I am really not sure why or how to help, but it reads a little strangely. Maybe:
```suggestion
Include documentation about your PR
```

""""""""""""""""""""""""""""""""

> Now, let users know about your new feature by adding it to the `WarpX documentation <https://warpx.readthedocs.io>`_.

Reviewer: ```suggestion
Now, let users know about your new feature by adding it's description to the `WarpX documentation <https://warpx.readthedocs.io>`_.
And correct/add any `doxygen` information over any function you might have changed/added, following the `<Manual http://www.doxygen.nl/manual/docblocks.html>`_
```

Author: thx, adopted :+1:

> Our documentation uses `Sphinx <http://www.sphinx-doc.org/en/master/usage/quickstart.html>`_, and it is located in ``Docs/``.

Reviewer: Just FYI, actually, regarding the sphinx documentation.
I can't see the built html files correctly (format wise) on the safari browser (I know my Catalina OSX... annoying us again!).
I installed sphinx via `spack`, according to <https://spack.readthedocs.io/en/latest/contribution_guide.html>:
```zsh
$ spack install py-sphinx
$ spack install py-sphinxcontrib-programoutput
$ spack install py-sphinx-rtd-theme
$ spack install graphviz
$ spack install git
$ spack install mercurial
$ spack install subversion
$ spack activate py-sphinx
$ spack activate py-sphinx-rtd-theme
$ spack activate py-sphinxcontrib-programoutput
```
But firefox opens those files very well!

Author: Yes, or here: https://openpmd-api.readthedocs.io/en/0.11.1-alpha/dev/sphinx.html

Author: Can maybe be moved in its own section in a follow-up.

For instance, if you introduce a new runtime parameter in the input file, you can add it to :ref:`Docs/source/running_cpp/parameters.rst <running-cpp-parameters>`.
If Sphinx is installed on your computer, you should be able to generate the html documentation with

.. code-block:: sh

   make html

in ``Docs/``. Then open ``Docs/build/html/index.html`` with your favorite web browser and look
for your changes.

Once your code is ready with documentation and automated test, congratulations!
You can create the PR (or remove the ``[WIP]`` tag if you already created it).
Reviewers will interact with you if they have comments/questions.

Style and conventions
---------------------

- For indentation, WarpX uses four spaces (no tabs)

> - The number of characters per line should be <100

Reviewer: ```suggestion
- The number of characters per line should be <100 (except in documentation files, where each line should be the length of a full sentence, to allow for easy changes/corrections in GitHub)
```

Author: Urgh, I had that in here but accidentally removed all changes on the first go before commit xD

> - To define a function , for e.g., ``myfunction()`` use a space between the name of the function and the paranthesis - ``myfunction ()``.

Reviewer: ```suggestion
- Space before and after you use `=` to define variables
- To define a function , for e.g., ``myfunction()`` use a space between the name of the function and the paranthesis - ``myfunction ()``.
```

Author: right, and so much more :D

  To call the function, the space is not required, i.e., just use ``myfunction()``.

>   - The reason this is beneficial is that when we do a ``git grep`` to search for ``myfunction ()``, we can clearly see the locations where ``myfunction ()`` is defined and where ``myfunction()`` is called.

Reviewer: This is super useful to know @ax3l !! 👍

Author: Oh that's all @WeiqunZhang's great life hack, I just reformatted this for inclusion in the manual ;)

  - Also, using ``git grep "myfunction ()"`` searches for files only in the git repo, which is more efficient compared to the ``grep "myfunction ()"`` command that searches through all the files in a directory, including plotfiles for example.
- It is recommended that style changes are not included in the PR where new code is added.
  This is to avoid any errors that may be introduced in a PR just to do style change.
- Some text editors automatically modify the files you open.

>   We recommend to turn on to remove trailing spaces and replace Tabs with 4 spaces.

Reviewer: I would put these 2 lines as a note afterwards or before, if you choose to mention the `.github/workflow` files.

Author: moved down
